src/ngraph_encapsulate_op.cc

```diff
@@ -123,24 +123,23 @@ class NGraphEncapsulateOp : public OpKernel {
     }
 
     // Create the backend
-    if (s_ng_backend == nullptr) {
-      mutex_lock l(s_ng_backend_mutex);
-
-      if (s_ng_backend == nullptr) {
-        const char* ng_backend_env_value = std::getenv("NGRAPH_TF_BACKEND");
-        if (ng_backend_env_value != nullptr) {
-          s_ng_backend_name = std::string(ng_backend_env_value);
-          if (s_ng_backend_name.empty()) {
-            s_ng_backend_name = "CPU";
-          }
-        } else {
-          s_ng_backend_name = "CPU";
-        }
-        s_ng_backend = ng::runtime::Backend::create(s_ng_backend_name);
-        OP_REQUIRES(ctx, s_ng_backend != nullptr,
-                    errors::InvalidArgument("Cannot create nGraph backend"));
+    if (auto ptr = s_ng_backend_wptr.lock()) {
+        m_ng_backend = ptr;
+        return;
+    }
+    const char* ng_backend_env_value = std::getenv("NGRAPH_TF_BACKEND");
+    if (ng_backend_env_value != nullptr) {
+      s_ng_backend_name = std::string(ng_backend_env_value);
+      if (s_ng_backend_name.empty()) {
+        s_ng_backend_name = "CPU";
       }
+    } else {
+      s_ng_backend_name = "CPU";
     }
+    m_ng_backend = ng::runtime::Backend::create(s_ng_backend_name);
+    OP_REQUIRES(ctx, m_ng_backend != nullptr,
+                errors::InvalidArgument("Cannot create nGraph backend"));
+    s_ng_backend_wptr = m_ng_backend;
```

Comment:
I think there's a possibility of a race condition with the initialization of `s_ng_backend_wptr` here:

```
t1 calls Backend::create and assigns m_ng_backend
t2 calls Backend::create and assigns m_ng_backend
t2 assigns s_ng_backend_wptr
t1 assigns s_ng_backend_wptr
```

Reply:
yes, you are right, I was think it will be fine since the assign for weak_ptr is atomic, but I realize the it will be create and assign multiple times. Let me add a mutex.

Reply:
fixed.